File: src/test/java/com/hileco/cortex/fuzzer/ProgramGeneratorFuzzTest.java
package com.hileco.cortex.fuzzer;

import com.hileco.cortex.vm.VirtualMachine;
import com.hileco.cortex.vm.Program;
import com.hileco.cortex.vm.ProgramContext;
import com.hileco.cortex.instructions.ProgramException;
import com.hileco.cortex.instructions.ProgramRunner;
import org.junit.Assert;
import org.junit.Test;

import java.math.BigInteger;
import java.util.ArrayList;

public class ProgramGeneratorFuzzTest {

    private static final int LIMIT_RUNS = 100_000;

    @Test
    public void fuzzTestGenerator() {
        var seed = System.currentTimeMillis() * LIMIT_RUNS;
        long exceptions = 0;
        long runs = 0;
        while (runs < LIMIT_RUNS) {
            var programGenerator = new ProgramGenerator();
            var generated = programGenerator.generate(seed + runs);
            var programAddresses = generated.keySet();
            for (var programAddress : programAddresses) {
                runs++;
                try {
                    var caller = new Program(BigInteger.ZERO, new ArrayList<>());
                    var program = generated.get(programAddress);
                    var callerContext = new ProgramContext(caller);
                    var programContext = new ProgramContext(program);
                    var processContext = new VirtualMachine(callerContext, programContext);
                    var programRunner = new ProgramRunner(processContext);
                    programRunner.run();
                } catch (ProgramException e) {
                    exceptions++;
                }
            }
        }
        var ratio = ((double) exceptions) / ((double) runs);
        Assert.assertFalse(String.format("Too many ProgramExceptions per generated program: %f (seed: %d)", ratio, seed), ratio > 0.5);
    }
}

File: src/test/java/com/hileco/cortex/fuzzer/ProgramGeneratorFuzzTest.kt
package com.hileco.cortex.fuzzer

import com.hileco.cortex.documentation.Documentation
import com.hileco.cortex.instructions.ProgramException
import com.hileco.cortex.instructions.ProgramRunner
import com.hileco.cortex.vm.concrete.Program
import com.hileco.cortex.vm.concrete.ProgramContext
import com.hileco.cortex.vm.concrete.VirtualMachine
import org.junit.Assert
import org.junit.Test

class ProgramGeneratorFuzzTest {
    @Test
    fun fuzzTestGenerator() {
        val seed = System.currentTimeMillis() * LIMIT_RUNS
        var exceptions: Long = 0
        var runs: Long = 0
        while (runs < LIMIT_RUNS) {
            val programGenerator = ProgramGenerator()
            val generated = programGenerator.generate(seed + runs)
            val programAddresses = generated.keySet()
            for (programAddress in programAddresses) {
                runs++
                try {
                    val caller = Program(listOf())
                    val program = generated[programAddress]!!
                    val callerContext = ProgramContext(caller)
                    val programContext = ProgramContext(program)
                    val virtualMachine = VirtualMachine(callerContext, programContext)
                    val programRunner = ProgramRunner(virtualMachine)
                    programRunner.run()
                } catch (e: ProgramException) {
                    exceptions++
                }

            }
        }
        val ratio = exceptions.toDouble() / runs.toDouble()
        Assert.assertFalse("Too many ProgramExceptions per generated program: $ratio (seed: $seed)", ratio > 0.5)
    }

    @Test
    fun testFuzzerSample() {
        val programGenerator = ProgramGenerator()
        val generated = programGenerator.generate(0)
        val first = generated.keySet().first()
        val program = generated[first]
        Documentation.of("fuzzer/sample").source(program!!.instructions)
        Assert.assertTrue(!program.instructions.isEmpty())
    }

    companion object {
        private const val LIMIT_RUNS = 1000
    }
}
